server/src/main/java/org/elasticsearch/action/support/ParameterizedTransportAction.java

```diff
@@ -0,0 +1,190 @@
+        handler.invoke(task, request, listener);
+    }
+
+    private Invoker<Request, Response> resolveHandler(Services services) {
+        final List<Method> methods = Arrays.stream(this.getClass().getMethods())
+            .filter(method -> method.isAnnotationPresent(ActionHandler.class))
+            .toList();
+        final Method executeMethod = switch (methods.size()) {
+            case 1 -> methods.getFirst();
+            case 0 -> throw new IllegalStateException(
+                "cannot find @" + ActionHandler.class.getSimpleName() + " annotated method in [" + getClass() + "]"
+            );
+            default -> throw new IllegalStateException(
+                "found multiple ["
+                    + methods.size()
+                    + "] @"
+                    + ActionHandler.class.getSimpleName()
+                    + " annotated methods in ["
+                    + getClass()
+                    + "] : "
+                    + methods
+            );
+        };
+
+        final Class<?>[] parameterTypes = executeMethod.getParameterTypes();
+        final List<ParameterSupplier<Request, Response>> parameterSuppliers = new ArrayList<>(parameterTypes.length);
+
+        for (int i = 0; i < parameterTypes.length; i++) {
+            final Class<?> type = parameterTypes[i];
+            try {
+                ParameterSupplier<Request, Response> param = resolveParameter(type, services);
+                parameterSuppliers.add(param);
+            } catch (IllegalArgumentException e) {
+                throw new ElasticsearchException(
+                    "cannot provide parameter of type [{}] for parameter #{} to {}",
+                    e,
+                    type,
+                    i + 1,
+                    executeMethod
+                );
+            }
+        }
+
+        final MethodHandle handle;
+        try {
+            handle = MethodHandles.lookup().unreflect(executeMethod);
+        } catch (IllegalAccessException e) {
+            throw new ElasticsearchException("cannot access @{} method [{}]", e, ActionHandler.class.getSimpleName(), executeMethod);
+        }
+        return buildInvoker(parameterSuppliers, handle, executeMethod);
+    }
+
+    @SuppressForbidden(reason = "Need to use MethodHandle.invoke because we don't know the argument types at compile time")
+    private Invoker<Request, Response> buildInvoker(
+        List<ParameterSupplier<Request, Response>> parameterSuppliers,
+        MethodHandle handle,
+        Method executeMethod
+    ) {
+        return (task, request, listener) -> {
+            Object[] args = new Object[parameterSuppliers.size() + 1];
+            args[0] = this;
+            for (int i = 1; i < args.length; i++) {
+                ParameterSupplier<Request, Response> ps = parameterSuppliers.get(i - 1);
+                args[i] = ps.resolveParameter(task, request, listener);
+            }
+            try {
+                handle.invokeWithArguments(args);
```

Comment:
If we wanted we could use ASM to build something that does more efficient invocation (which wouldn't even need to be a method handle), but I think the performance on this should be satisfactory for most cases.